### DISCUSSION TITLE
Adding `align_corners` to `jax.image.resize`

### DISCUSSION THREAD

**Discussion by younesbelkada** [google/flax#2211] [Q&A] [answered]
Hi all! 

I am currently trying to convert DPT (Dense Prediction Transformers) model on Flax to add it on `transformers` library 🚀 
 https://github.com/huggingface/transformers/pull/17779
The model heavily relies on [`torch.nn.Upsample`](https://github.com/huggingface/transformers/blob/da27c4b398e607c161451f335367ad666de08497/src/transformers/models/dpt/modeling_dpt.py#L839) and [interpolation operations](https://github.com/huggingface/transformers/blob/da27c4b398e607c161451f335367ad666de08497/src/transformers/models/dpt/modeling_dpt.py#L608) to upsample the intermediate output and use them in a skip-connection manner. 
In my Flax implementation I use `jax.image.resize` to interpolate the intermediate output + final results and I wrapped it in a `Module` as follows:
```
class FlaxDPTUpsample(nn.Module):
    scale: int = 2
    method: str = "bilinear"

    def setup(self):
        pass

    def __call__(self, x, output_size=None):
        if output_size is None:
            output_size = x.shape
            output_size = (output_size[0], output_size[1] * self.scale, output_size[2] * self.scale, output_size[3])
        return jax.image.resize(x, output_size, method=self.method)
```
The problem now is that I do not get exactly the same results as the PyTorch implementation, and I suspect that it is related to the fact that in the PyTorch implementation, the model uses `align_corners=True` on the interpolation+upsampling operations which obviously do not give the same result as when using `align_corners=False`. I was wondering if you are planning to add this argument to the current implementations of interpolations? Or is there any hack / way to implement this (I could also open a PR?)
I would also appreciate any help / tips regarding my WIP implementation of DPT - for example I am not sure whether to use `jax.image.resize` or `jax.image.scale_and_translate` in my case, also if it is correct to use these functions as a 2d interpolation (ie an equivalent of `torch.nn.functional.interpolate` )
Thank you very much in advance 🙏 

cc @marcvanzee @cgarciae

**Comment by cgarciae** [comment, marked as answer]
Hey @younesbelkada! I did this simple test and can verify that `jax.image.resize` aligns corners by default:

```python
import jax.numpy as jnp
import jax

input_arr = jnp.array([
    [0, 1, 2, 3, 4],
    [5, 6, 7, 8, 9],
])


output = jax.image.resize(
    image=input_arr,
    shape=(3,6),
    method="bilinear")

print(output)
# [[0.        0.75      1.5833333 2.4166665 3.25      4.       ]
# [2.5       3.25      4.0833335 4.9166665 5.75      6.5      ]
# [5.        5.75      6.5833335 7.4166665 8.25      9.       ]]
```
Two suggestions:
1. Create/share a unit test comparing with the target implementation. That way we can help you a bit more.
2. If the problem is indeed in resize, create an issue in [google/jax](https://github.com/google/jax/issues/new/choose)

I'll be converting this to a discussion.

**Reply by younesbelkada** [reply]
Hey @cgarciae ,

Thank you very much for your reply!
Here is a script I have used to confirm that `jax.image.resize` does not reproduce the same output as `torch.nn.Upsample(align_corners=True)`:

```
import torch
import torch.nn as nn
import jax.numpy as jnp
import jax
import numpy as np

input_arr = jnp.array([
    [0, 1, 2, 3, 4],
    [5, 6, 7, 8, 9],
]) / 9

torch_input_arr = torch.from_numpy(np.array(input_arr)).unsqueeze(0).unsqueeze(0)
upsample_with_align_corners = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)
upsample_without_align_corners = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=False)

output = jax.image.resize(
    image=input_arr,
    shape=(4,10),
    method="bilinear")

output_torch_upsample_with_align_corners = upsample_with_align_corners(torch_input_arr)
output_torch_upsample_without_align_corners = upsample_without_align_corners(torch_input_arr)

print("jax: ", output)
print("torch upsample align corner:", output_torch_upsample_with_align_corners)
print("torch upsample without align corners",output_torch_upsample_without_align_corners)
```
from the prints I get: 
```
# jax default output
DeviceArray([[0.        , 0.02777778, 0.08333334, 0.1388889 , 0.19444445,
              0.25      , 0.30555555, 0.3611111 , 0.4166667 , 0.44444445],
             [0.1388889 , 0.16666667, 0.22222222, 0.2777778 , 0.33333334,
              0.3888889 , 0.44444442, 0.5       , 0.5555556 , 0.5833334 ],
             [0.4166667 , 0.44444448, 0.5       , 0.5555555 , 0.6111111 ,
              0.6666667 , 0.7222222 , 0.7777778 , 0.8333333 , 0.8611111 ],
             [0.5555556 , 0.5833334 , 0.6388889 , 0.6944444 , 0.75      ,
              0.8055556 , 0.8611111 , 0.9166667 , 0.9722222 , 1.        ]],            dtype=float32)
# torch with align corner
tensor([[[[0.0000, 0.0494, 0.0988, 0.1481, 0.1975, 0.2469, 0.2963, 0.3457,
           0.3951, 0.4444],
          [0.1852, 0.2346, 0.2840, 0.3333, 0.3827, 0.4321, 0.4815, 0.5309,
           0.5802, 0.6296],
          [0.3704, 0.4198, 0.4691, 0.5185, 0.5679, 0.6173, 0.6667, 0.7160,
           0.7654, 0.8148],
          [0.5556, 0.6049, 0.6543, 0.7037, 0.7531, 0.8025, 0.8519, 0.9012,
           0.9506, 1.0000]]]])
# torch without align corner
tensor([[[[0.0000, 0.0278, 0.0833, 0.1389, 0.1944, 0.2500, 0.3056, 0.3611,
           0.4167, 0.4444],
          [0.1389, 0.1667, 0.2222, 0.2778, 0.3333, 0.3889, 0.4444, 0.5000,
           0.5556, 0.5833],
          [0.4167, 0.4444, 0.5000, 0.5556, 0.6111, 0.6667, 0.7222, 0.7778,
           0.8333, 0.8611],
          [0.5556, 0.5833, 0.6389, 0.6944, 0.7500, 0.8056, 0.8611, 0.9167,
           0.9722, 1.0000]]]])
```
From this observation it seems that `output_torch_upsample_without_align_corners` is closer to `output` than `output_torch_upsample_with_align_corners` => that is how I concluded that `jax.image.resize` is not behaving exactlty as `nn.Upsample(align_corners=True)`

**Reply by younesbelkada** [reply]
@cgarciae I think that you are right and in all the cases mentioned above it aligns the corners:
![image](https://user-images.githubusercontent.com/49240599/174779833-660b9cb4-f75d-488d-a22c-a186f2c212e3.png)
But somehow the results are different, do you have by chance any idea why we are not matching the results of `nn.Upsample(align_corners=True)` ?

**Reply by cgarciae** [reply]
I was playing with this, it seems `align_corners=True` doesn't mean "align the corners" but rather "sample with equal spacing" or something like that. I would highly encourage you to post an issue in the jax repo, this issue seems relevant to JAX in general.

**Reply by younesbelkada** [reply]
Great ! Will post an issue now

**Reply by younesbelkada** [reply]
Just posted here: https://github.com/google/flax/issues/2214 !
